django_toolkit/fallbacks/circuit_breaker.py

```diff
@@ -34,21 +34,25 @@ def total_failures(self):
         return self.cache.get(self.failure_cache_key) or 0
 
     def open_circuit(self):
+        self.cache.set(self.circuit_cache_key, True, self.circuit_timeout)
+
         logger.critical(
             'Open circuit for {failure_cache_key} {cicuit_cache_key}'.format(
                 failure_cache_key=self.failure_cache_key,
                 cicuit_cache_key=self.circuit_cache_key
             )
         )
 
-        self.cache.set(self.circuit_cache_key, True, self.circuit_timeout)
-
     def __enter__(self):
         if self.is_circuit_open:
             raise self.max_failure_exception
 
+        return self
+
     def __exit__(self, exc_type, exc_value, traceback):
         if exc_type in self.catch_exceptions:
+            if self.is_circuit_open:
+                raise self.max_failure_exception
 
             self._increase_failure_count()
 
```

Comment:
This code is executed only when the code inside the context failed and raised an exception listed in `catch_exceptions` tuple. When the code inside the context succeeds, this new code will not be called.

Moving to the below `if` would increment the counter, even when the circuit was already open. Is this the expected behaviour?